### DISCUSSION TITLE
Using non linear elasticity in Drucker Prager Cap Model for Code-Aster

### DISCUSSION THREAD

**Discussion by sameenmustafa** [thelfer/tfel#455] [Q&A]
Dear Dr. Helfer,
I writing here as I have been trying to implement the DPC model created by you in MFRONT. I have succeeded in compiling and implementation in code-aster. However, I have to simulate powder compaction using DPC model, for this, I have with me the Elastic Modulus and Posisson's coefficient as a function of pressure. This comes from the mechanical testing and calibration of the powder itself.
I have noticed that changing these properties in Code-aster doesn't affect the results as the simulation is being governed by the mfront library. To verify this, I have simulated using both the linear elastic properties as well as pressure dependent properties. Thus, I am writing my problems below:
1. How can I implement the varying E and nu in the Drucker Prager Cap Model code. I don't have a proper function, but I have the values of E and nu corresponding to different pressures (SIGM). To better explain this, I have the density of the powder as a function of the applied pressure but, I cannot implement the density as a function of pressure in code-aster as it only accepts density variation based on the co-ordinates. In the end, every varying input parameter needs to be inserted as a function of pressure. I am also attaching the .comm files of the simulation results for your reference.
2. While in powder compaction, in consideration with the DPC model, other parameters also vary as a function of density and in turn, as a function of pressure. For example, in my case, I would be having d, beta and R as constant while pa and pb needs to be varied in accordance with the callibration data. At this point I would also like to ask how can I use pa and pb as a user defined function based on pressure.
3. I would also like to implement a modified version of the DPC model, which includes a transition surface to smoothen the transition from the shear line to the cap. Since you are the main developer of the DPC model, can you please help me in defining the modified DPC model in MFRONT?

I am attaching here different files to refer to the simulations I have carried out so far. I am also new to MFRONT and Code-Aster as well, and after reading several discussions and watching different videos by Mr Cyprian and Dr Nicolo Grilli, I am writing here.

Thanks and Regards,
Yours Sincerely,
Sameen Mustafa.
[DPC.zip](https://github.com/thelfer/tfel/files/13528058/DPC.zip)

**Comment by thelfer** [comment]
@sameenmustafa Thanks for you interest to MFront.

This matter already have been discussed on this forum: https://github.com/thelfer/tfel/issues/391

You may want to read the following tutorials:

- https://thelfer.github.io/MFrontGallery/web/CrushedSaltKorthausBehaviour.html
- https://thelfer.github.io/tfel/web/MohrCoulomb.html

**Reply by sameenmustafa** [reply]
@thelfer 
Sir,
After going through the documentation I tried to implement  computestress and tangent operators but I am not able to compile it. Further, the page containing the constitutive equation and their implementation in mfront is not accessible anymore.
[DPCnew.txt](https://github.com/thelfer/tfel/files/13998713/DPCnew.txt)
I am attaching the modified code here. I have also made same modifications as I have to use the model in compressive state.  Please have a look. Could you please help me in moving head as I am new to MFRONT.

Kind regards,
Sameen

**Reply by thelfer** [reply]
@sameenmustafa The sourceforge website has been closed. It was deprecated in favour of the github website: https://thelfer.github.io/tfel/web/index.html

**Reply by thelfer** [reply]
@sameenmustafa Please find attach a compiling version. 
[DPCnew.txt](https://github.com/thelfer/tfel/files/14006869/DPCnew.txt)

This file does compile, but obsviously the computation of the stress and tangent operator are wrong.

**Reply by sameenmustafa** [reply]
@thelfer 
Dear Sir,
I have made a few changes by going through different forums and discussions advised by you previously. Now the code compiles without and warning or errors, but I am still doubtful about the computation of the stress and tangent operator.
Please have a look at the file attached below:
[DPC-edited (2).txt](https://github.com/thelfer/tfel/files/14269389/DPC-edited.2.txt)
Could you please guide me as to how can I do the two computations correctly.

Kind Regards,
Sameen

**Reply by thelfer** [reply]
The implementation is basically what I would have written without the `StandardElasticity` brick, although somethings are obviously wrong:

- the computation of the prediction of the stress
- the computation of the stress

If you want to compute the stress incrementally, you shall proceed as follows:

```c++
@LocalVariables StressStensor sig_bts;

@InitiLocalVariables{
  sig_bts = sig;
  StressStensor sigel = sig_bts + (lambda * trace(deto) * Stensor::Id() + 2 * mu * (deto));
  ...
}

...

@ComputeStress {
  sig = sig_bts + (lambda * trace(deel) * Stensor::Id() + 2 * mu * (deel));
}
```

**Reply by sameenmustafa** [reply]
@thelfer 
Sir,
Could you now please check if the implementation is fine..
I am attaching the modified file below:
[DPC-edited.2.txt](https://github.com/thelfer/tfel/files/14405117/DPC-edited.2.txt)

It compiles but gives the error that variable sigel initialised but not used.

Regards,
Sameen

**Reply by thelfer** [reply]
Just comment line 49. You indeed don't use the variable `sigel` in `@InitLocalVariables`

**Reply by sameenmustafa** [reply]
@thelfer 
Thankyou.
Is everything else fine? Tangent operator as well?
Or do I need to also add the part of the stress coming from the Cap criterion to the computestress block as well?

**Reply by thelfer** [reply]
> Is everything else fine? Tangent operator as well?

I think so

**Reply by sameenmustafa** [reply]
@thelfer
Sir,
I was wondering if I could use the youngs modulus as a function of pressure. I tried formulate the code below:
(@)Input rho;
rho.setGlossaryName("MassDensity");
@PhysicalBounds rho in [0:*[;

@MaterialProperty stress young;
@PhysicalBounds young in [0:*[;
young.setGlossaryName("YoungModulus");

(@)Function{
young =(-28e9+10120e6*rho)*exp(rho/6800)**6);
}

Would this approach work? If yes, can I input the density values in the material model in mfront code using code aster as follows:
MFRONT_FO = _F( LIST_COEF = (rho))

**Reply by thelfer** [reply]
Hi @sameenmustafa,

To be honest, I don't know if you could do this in `code_aster` in practice (I am not a reguler `code_aster` user).

Personaly, I would do it all in `mfront`, but it surely requires more work.

However, if you consider that the mass density is constant over the time step, you could simply update Young's modulus in `@InitializeLocalVariables` using the value of `rho` at the beginning of the time step.  However, I don't know if this is a reasonable approximation.

**Reply by sameenmustafa** [reply]
@thelfer
The density would be constant at each time-step, where 100 timesteps will have 100 values of densities different from each other. But the density depends upon the imposed displacement which would need to be input. This seems to be a bit complex if I use InitialiseLocalVariables.
Regarding doing it all in MFRONT, could you please suggest some resources.

Many Thanks.

**Reply by thelfer** [reply]
`@InitializeLocalVariables` is called at every time step at each integration point, so I don't get your point

**Reply by Asocie13** [reply]
Density depend on the volume/volumetric strain not the displacement. You can specify density such as an AuxiliaryStateVariable and update it in the section @UpdateAuxiliaryStateVariables.

**Reply by sameenmustafa** [reply]
@thelfer 
I am working on powder compaction. For now I Plan to input the displacement at each time step, like 1mm at 0.1 sec. So the instantaneous density at a particular time step would be (mass/instantaneous volume), where the mass is constant throughout. But I am using Salome-Meca to impose these displacements. I am able to compile mfront code and generate the library, then I pass the required material properties using MFRONT module in salome itself.
This would work fine if the E and nu were a constant value, but in the case of powder these value vary with density. I am a bit confused as how to proceed now.

**Reply by thelfer** [reply]
As stated by @Asocie13, the evolution of the density is known locally based on either the total strain or the plastic strain depending on the hypothesis that you want to use. See: https://thelfer.github.io/MFrontGallery/web/CrushedSaltKorthausBehaviour.html for example, but there are other resources on the TFEL website: https://thelfer.github.io/tfel/web/StandardElastoViscoPlasticityBrick-PorousPlasticity.html

**Reply by Asocie13** [reply]
Thomas also specify how change the DPC mfront file in https://github.com/thelfer/tfel/blob/master/docs/web/drucker-prager-cap.md

With Salome, you can manage how to model powder compaction (imposed displacement or pressure, matrix-powder contact) and mfront file models the local material behavior. Depending on the local compaction you will model the volumetric strain variation and thus the density variation.

**Reply by sameenmustafa** [reply]
@thelfer
Hello Sir,
I was simulating on Salome using the DPC Mfront code from the repository, I am getting an error which I tried to resolve but it does not work. Could you help in this regard:
EXCEPTION> <CALCUL_46>                                                           !
   !                                                                                   !
   !  Il y a une incohérence entre les paramètres matériaux et la loi de comportement. !
   !  On ne trouve pas les paramètres matériaux associés à la relation MFRONT          !
   !                                                                                   !
   !  pour la maille M3433.                                                            !
   !                                                                                   !
   !  Conseils :                                                                       !
   !     Vérifiez la mise en données de la commande DEFI_MATERIAU : il faut définir    !
   !     les paramètres matériau pour la relation utilisée.                            !
   !     Vérifiez que la maille M3433 est bien affectée par le matériau.               !
   !                                                                                   !
   ! --------------------------------------------                                      !
   ! Contexte du message :                                                             !
   !    Option         : RIGI_MECA_TANG                                                !
   !    Type d'élément : MECA_TETRA4                                                   !
   !    Maillage       : mesh                                                          !
   !    Maille         : M3433                                                         !
   !    Type de maille : TETRA4                                                        !
   !    Cette maille appartient aux groupes de mailles suivants :                      !
   !                                                                                   !
   !    Position du centre de gravité de la maille :                                   !
   !       x=0.003428 y=0.000398 z=0.009936                                            !
   
   DC.py : ERREUR WITH THE EXECUTION - INTERRUPTION
>> JDC.py: DEBUT CR of execution of JDC in MIXTE
   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
   ! <S> Exception user raised but not interceptee.                           !
   ! The bases are fermees.                                                   !
   ! Type of the exception: error                                             !
   !                                                                          !
   !  Il y a une incohérence entre les paramètres matériaux et la loi de      !
   ! comportement.                                                            !
   !  On ne trouve pas les paramètres matériaux associés à la relation MFRONT !
   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
fine CR of execution of JDC in MIXTE

**Reply by thelfer** [reply]
I think that the `DEFI_MATERIAU` command has an `MFRONT` keyword to define the properties associated with the material properties.

I don't know with version of aster you used, but for old versions, there is a long standing bug in the `code_aster` interface that requires the `MFront` behaviours to have two material properties at least, so you may have to add:

~~~~
@MaterialProperty real dummy1, dummy2;
~~~~

**Comment by sameenmustafa** [comment]
@thelfer
Sir, Could you please have a look at the last message...

**Reply by thelfer** [reply]
@sameenmustafa, I don't know what happened, I did make comments. Let me see again.

**Reply by sameenmustafa** [reply]
@thelfer 
Thankyou for replying. I had not recieved the previous message.
